libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/location/replay/ReplayRouteLocationEngine.java

```diff
@@ -0,0 +1,168 @@
+    converter = new ReplayRouteLocationConverter(route);
+    mockedLocations = converter.toLocations();
+    dispatcher = obtainDispatcher();
+    dispatcher.run();
+    int mockedPoints = mockedLocations.size();
+    if (mockedPoints <= 5) {
+      handler.postDelayed(this, 1000);
+    } else {
+      handler.postDelayed(this, (mockedPoints - 5) * 1000);
+    }
+  }
+
+  public void assignLastLocation(Point currentPosition) {
+    lastLocation.setLongitude(currentPosition.longitude());
+    lastLocation.setLatitude(currentPosition.latitude());
+  }
+
+  @SuppressLint("MissingPermission")
+  public void moveTo(Point point) {
+    Location lastLocation = getLastLocation();
+    if (lastLocation == null) {
+      return;
+    }
+
+    List<Point> pointList = new ArrayList<>();
+    pointList.add(Point.fromLngLat(lastLocation.getLongitude(), lastLocation.getLatitude()));
+    pointList.add(point);
+
+    LineString route = LineString.fromLngLats(pointList);
+
+    handler.removeCallbacks(this);
+    mockedLocations = converter.calculateMockLocations(converter.sliceRoute(route));
+    dispatcher = obtainDispatcher();
+    dispatcher.run();
+  }
+
+  @Override
+  public void run() {
+    if (mockedLocations.isEmpty()) {
+      handler.removeCallbacks(this);
+    } else {
+      List<Location> nextMockedLocations = converter.toLocations();
+      dispatcher.add(nextMockedLocations);
+      mockedLocations.addAll(nextMockedLocations);
+      int currentMockedPoints = mockedLocations.size();
+      if (currentMockedPoints <= 5) {
+        handler.postDelayed(this, 1000);
+      } else {
+        handler.postDelayed(this, (currentMockedPoints - 5) * 1000);
+      }
+    }
+  }
+
+  /**
+   * Connect all the location listeners.
+   */
+  @Override
+  public void activate() {
+    for (LocationEngineListener listener : locationListeners) {
+      listener.onConnected();
+    }
+  }
+
+  @Override
+  public void deactivate() {
+    if (dispatcher != null) {
+      dispatcher.stop();
+    }
+    handler.removeCallbacks(this);
+  }
+
+  /**
+   * While the {@link ReplayRouteLocationEngine} is in use, you are always connected to it.
+   *
+   * @return true.
+   */
+  @Override
+  public boolean isConnected() {
+    return true;
+  }
+
+  @SuppressLint("MissingPermission")
+  @Override
+  @Nullable
+  public Location getLastLocation() {
+    return lastLocation;
+  }
+
+  /**
+   * Nothing needs to happen here since we are mocking the user location along a route.
+   */
+  @Override
+  public void requestLocationUpdates() {
+
+  }
+
+  /**
+   * Removes location updates for the LocationListener.
+   */
+  @Override
+  public void removeLocationUpdates() {
+    for (LocationEngineListener listener : locationListeners) {
+      locationListeners.remove(listener);
+    }
+    if (dispatcher != null) {
+      dispatcher.removeReplayLocationListener(replayLocationListener);
+    }
+  }
+
+  @Override
+  public Type obtainType() {
+    return Type.MOCK;
```

Comment:
We should update this constant to reflect the new class naming

Reply:
Unfortunately, we can't. This is a [`Type`](https://github.com/mapbox/mapbox-events-android/blob/5b159c98f0eb03de9a54ae88443198156bbcd951/liblocation/src/main/java/com/mapbox/android/core/location/LocationEngine.java#L18-L20) exposed upstream in [`mapbox-events-android`](https://github.com/mapbox/mapbox-events-android). BTW at the end of the day this is a "mock" location engine anyways, right? For now, I think is fine, we can cut a ticket later upstream if needed. What do you think?

Reply:
https://github.com/mapbox/mapbox-events-android/pull/215

Reply:
> https://github.com/mapbox/mapbox-events-android/pull/215

Follow up 👉 https://github.com/mapbox/mapbox-events-android/pull/215#issuecomment-420947181